src/ILCompiler.Compiler/src/Compiler/TypePreinit.cs

```diff
@@ -0,0 +1,1964 @@
+                            int numParams = methodSig.Length + paramOffset;
+
+                            TypeDesc owningType = method.OwningType;
+                            if (!_compilationGroup.CanInline(methodIL.OwningMethod, method)
+                                || (owningType.HasStaticConstructor
+                                    && owningType != methodIL.OwningMethod.OwningType
+                                    && !((MetadataType)owningType).IsBeforeFieldInit))
+                            {
+                                return Status.Fail(methodIL.OwningMethod, opcode);
+                            }
+
+                            Value[] methodParams = new Value[numParams];
+                            for (int i = numParams - 1; i >= 0; i--)
+                            {
+                                methodParams[i] = stack.PopIntoLocation(GetArgType(method, i));
+                            }
+
+                            Value retVal;
+                            if (!method.IsIntrinsic || !TryHandleIntrinsicCall(method, methodParams, out retVal))
+                            {
+                                recursionProtect ??= new Stack<MethodDesc>();
+                                recursionProtect.Push(methodIL.OwningMethod);
+                                Status callResult = TryScanMethod(method, methodParams, recursionProtect, out retVal);
+                                if (!callResult.IsSuccessful)
+                                {
+                                    recursionProtect.Pop();
+                                    return callResult;
+                                }
+                                recursionProtect.Pop();
+                            }
+
+                            if (!methodSig.ReturnType.IsVoid)
+                                stack.PushFromLocation(methodSig.ReturnType, retVal);
+                        }
+                        break;
+
+                    case ILOpcode.newobj:
+                        {
+                            MethodDesc ctor = (MethodDesc)methodIL.GetObject(reader.ReadILToken());
+                            MethodSignature ctorSig = ctor.Signature;
+
+                            TypeDesc owningType = ctor.OwningType;
+                            if (!_compilationGroup.CanInline(methodIL.OwningMethod, ctor)
+                                || !_compilationGroup.ContainsType(owningType)
+                                || !owningType.IsDefType
+                                || owningType.HasFinalizer
+                                || ((DefType)owningType).ContainsGCPointers
+                                || (owningType.HasStaticConstructor
+                                    && owningType != methodIL.OwningMethod.OwningType
+                                    && !((MetadataType)owningType).IsBeforeFieldInit))
+                            {
+                                return Status.Fail(methodIL.OwningMethod, opcode);
```

Comment:
It may be a good idea to include a separate reason for each of the conditions above. As I have experimented with this, my object was not getting pre-initialized. I had to step through the code to figure out the condition that it was tripping.

Reply:
Yeah, I got lazy. I'll fix them up.